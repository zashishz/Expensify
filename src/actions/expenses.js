import uuid from 'uuid';
import database from '../firebase/firebase';

// Expenses Actions

export const addExpense = (expense) => ({
    type: "ADD_EXPENSE",
    expense
});

// Custom function to store Data to firebase - async-actions - react-thunk
export const startAddExpense = (expenseData = {}) => {
    return (dispatch, getState) => {
        const {

            description = '',
            note = '',
            amount = 0,
            createdAt = 0

        } = expenseData;

        const expense = { description, note, amount, createdAt };
        const uid = getState().auth.uid;

        return database.ref(`users/${uid}/expenses`).push(expense)
            .then((ref) => {
                dispatch(addExpense({ id: ref.key, ...expense }))
            });
    }
}

export const removeExpense = (id) => ({
    type: 'REMOVE_EXPENSE',
    id
});

export const startRemoveExpense = ( id ) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).remove().then(() => {
            dispatch(removeExpense(id));
        })
    }
}

export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
})

export const startEditExpense = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).update(updates).then(() => {
            dispatch(editExpense(id, updates));
        })
    }
}

export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

export const startSetExpenses = (expenseData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses`).once('value').then((snapshot) => {
            const expenses = [];
            snapshot.forEach((child) => {
                expenses.push({
                    id: child.key,
                    ...child.val()
                })
            })
            dispatch(setExpenses(expenses))
        })
    }
}