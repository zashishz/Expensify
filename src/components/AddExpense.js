import React from 'react';

const AddExpense = () => (
    <div>
        This is Add Expense!
    </div>
);

export default AddExpense;