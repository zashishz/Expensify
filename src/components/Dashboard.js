import React from 'react';

const Dashboard = () => (
    <div>
        This is Expesify Dashboard!
    </div>
);

export default Dashboard;