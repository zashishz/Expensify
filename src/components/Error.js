import React from 'react';

const Error = (props) => (
    <div>
        {props.error}
    </div>
)

export default Error;