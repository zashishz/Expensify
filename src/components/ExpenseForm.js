import React from 'react';

export default class ExpenseForm extends React.Component {
    render = () => (
        <div>
            <p>Expense Form</p>
        </div>
    )
}