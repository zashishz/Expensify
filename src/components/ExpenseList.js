import React from 'react';
import { connect } from 'react-redux';
import validateExpenses from '../selectors/expenses';
import ExpenseListItem from './ExpenseListItem';

export const ExpenseList = (props) => (
    <div>
        <h2>Expense List</h2>
        { 
            props.expenses.map((expense) => {
                return <ExpenseListItem key={expense.id} {...expense} />
            })
        }
    </div>
);

export default connect((state) => {
    return {
        expenses: validateExpenses(state.expenses, state.filters)
    }
})(ExpenseList);