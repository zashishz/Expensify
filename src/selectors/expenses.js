import moment from 'moment';

export default (expenses, { text, sortBy, startDate, endDate, createdAt }) => {

    const createdAtMoment = moment(createdAt);

    return expenses.filter((expense) => {
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day'): true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
    
        return startDateMatch && endDateMatch && textMatch;
    }).sort((a,b) => {
        if(sortBy == 'date') return a.createdAt < b.createdAt ? 1 : -1;
        else if(sortBy == 'amount') return a.amount < b.amount ? 1: -1;
    })
};