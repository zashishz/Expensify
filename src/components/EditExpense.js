import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {editExpense, startRemoveExpense} from '../actions/expenses';

export class EditExpense extends React.Component {
    
    editExpense = (expense) => {
        this.props.editExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    }

    startRemoveExpense = () => {
        this.props.startRemoveExpense(this.props.expense.id);
        this.props.history.push('/');
    }

    render() {
        return (
            <div>
                <ExpenseForm
                    expense = {this.props.expense}
                    onSubmit={ this.editExpense }
                />
                <button onClick={this.startRemoveExpense}>Remove</button>
            </div>
        )
    }
}


const mapDispatchToProps = (dispatch) => ({
    editExpense: (id, expense) => dispatch(editExpense(id, expense)),
    startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
})

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpense);