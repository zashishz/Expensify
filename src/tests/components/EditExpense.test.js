import React from 'react';
import { shallow } from 'enzyme';
import { EditExpense } from '../../components/EditExpense';
import expenses from '../fixtures/expenses';

let editExpense, startRemoveExpense, history, wrapper;

beforeEach(() => {
    editExpense = jest.fn();
    startRemoveExpense = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(<EditExpense editExpense={editExpense} startRemoveExpense={startRemoveExpense} history={history} expense={expenses[1]} />);
})

test('should render snapshot', () => {
    expect(wrapper).toMatchSnapshot();
})

test('should handle edit expense', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);
    expect(history.push).toHaveBeenCalledWith('/');
    expect(editExpense).toHaveBeenCalledWith(expenses[1].id, expenses[1]);
})

test('should handle edit expense', () => {
    wrapper.find('button').simulate('click');
    expect(history.push).toHaveBeenCalledWith('/');
    expect(startRemoveExpense).toHaveBeenCalledWith(expenses[1].id);
})