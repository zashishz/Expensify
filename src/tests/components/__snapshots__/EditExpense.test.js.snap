// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render snapshot 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 2200,
        "createdAt": -345600000,
        "description": "Rent",
        "id": "2",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
